lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -541,6 +543,29 @@ def confirm_namespace_exists
       @logger.info("Namespace #{@namespace} found")
     end
 
+    # make sure to never prune the ejson-keys secret
+    def check_applied_ejson_keys_secret
```

Comment:
The fact that EJSON specific details have leaked into this file seems like an abstraction issue.  Any reason we couldn't move this function into the ejson_secret_provisoner.rb and call it from https://github.com/Shopify/kubernetes-deploy/blob/master/lib/kubernetes-deploy/ejson_secret_provisioner.rb#L135

Reply:
`EjsonSecretProvisioner` only does work if a `secrets.ejson` [file exists in the deploy directory](https://github.com/Shopify/kubernetes-deploy/blob/master/lib/kubernetes-deploy/ejson_secret_provisioner.rb#L42). For our purposes, this isn't necessarily/routinely the case. E.g we use `ejson-keys` to decrypt secrets coming from other files.

I agree we're missing an abstraction layer, but I don't think `EjsonSecretProvisioner` is where this code needs to go, since it's more of a pre-flight check of a well-known resource that, under normal circumstances, _should not be in a position to be pruned_, as opposed to the general workflow of provisioning secrets via EJSON (the provisioner should solely be a consumer of `ejson-keys`, not mutate it in any way).

@KnVerey and I had a conversation regarding the bloat of `deploy_task.rb` in general, and it's certainly due for a refactor. Said that, since this issue deals with the behaviour of a kubernetes resource _at deploy time_, it should be the deploy task's responsibility to ensure correct behaviour. This speaks more towards abstracting away our preflight checks into a more cohesive validation/preadmission framework than anything else, IMO.

Reply:
The unique thing about this check is that it needs to run right here, not at the normal validation time, in order to preserve the ideal that our first few stages do not do any mutation. Therefore even when we have a proper shared validation abstraction, I kinda doubt this can be rolled into it. Maybe we should go ahead and create a new generic object now to encapsulate the concept of mutating preflight check or whatever we're calling it. After all this PR is the first time we do something like that.